File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> prefix(n, 1);
        vector<int> suffix(n, 1);

        for (int i = 0; i < n - 1; i++) {
            prefix[i + 1] = prefix[i] * nums[i];
        }
        for (int i = n - 1; i > 0; i--) {
            suffix[i - 1] = suffix[i] * nums[i];
        }

        // for (int i : prefix) cout << i << " ";
        // cout << endl;
        // for (int j : suffix) cout << j << " ";
        // cout << endl;

        vector<int> answer(n, 0);
        for (int i = 0; i < n; i++) {
            answer[i] = prefix[i] * suffix[i];
        }
        return answer;
    }
};

File: Data Structures & Algorithms/top-k-elements-in-list/submission-0.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        counter = dict()
        maxn = 1

        for i in nums:
            counter[i] = counter.get(i, 0) + 1
            maxn = max(maxn, counter[i])
        
        freq = [[] for _ in range(maxn + 1)]
        for el, cnt in counter.items():
            freq[cnt].append(el)
        
        answer = []
        for i in range(len(freq) - 1, 0, -1):
            for num in freq[i]:
                answer.append(num)
                if len(answer) == k:
                    return answer

